brainiak/utils/fmrisim.py

```diff
@@ -2046,38 +2247,345 @@ def _noise_dict_update(noise_dict):
+            fit the parameters to match the participant as best as possible.
+
+        fit_thresh : float
+            What proportion of the target parameter value is sufficient
+            error to warrant finishing fit search.
+
+        fit_delta : float
+            How much are the parameters attenuated during the fitting process,
+            in terms of the proportion of difference between the target
+            parameter and the actual parameter
+
+        iterations : int
+            The first element is how many steps of fitting the SFNR and SNR
+            values will be performed. Usually converges after < 5. The
+            second element is the number of iterations for the AR fitting.
+            This is much more time consuming (has to make a new timecourse
+            on each iteration) so be careful about setting this appropriately.
+
+        Returns
+        -------
+
+        noise : multidimensional array, float
+            Generates the noise volume given these parameters
+
+        """
+
+    # Pull out information that is needed
+    dim_tr = noise.shape
+    base = template * noise_dict['max_activity']
+    base = base.reshape(dim_tr[0], dim_tr[1], dim_tr[2], 1)
+    mean_signal = (base[mask > 0]).mean()
+    target_snr = noise_dict['snr']
+
+    # Iterate through different parameters to fit SNR and SFNR
+    spat_sd_orig = np.copy(spatial_sd)
+    iteration = 0
+    for iteration in list(range(iterations)):
+
+        # Calculate the new metrics
+        new_snr = _calc_snr(noise, mask)
+
+        # Calculate the difference between the real and simulated data
+        diff_snr = abs(new_snr - target_snr) / target_snr
+
+        # If the AR is sufficiently close then break the loop
+        if diff_snr < fit_thresh:
+            logger.info('Terminated SNR fit after ' + str(
+                iteration) + ' iterations.')
+            break
+
+        # Convert the SFNR and SNR
+        spat_sd_new = mean_signal / new_snr
+
+        # Update the variable
+        spatial_sd -= ((spat_sd_new - spat_sd_orig) * fit_delta)
+
+        # Prevent these going out of range
+        if spatial_sd < 0 or np.isnan(spatial_sd):
+            spatial_sd = 10e-3
+
+        # Set up the machine noise
+        noise_system = _generate_noise_system(dimensions_tr=dim_tr,
+                                              spatial_sd=spatial_sd,
+                                              temporal_sd=temporal_sd,
+                                              )
+
+        # Sum up the noise of the brain
+        noise = base + (noise_temporal * (1 - temporal_sd)) + noise_system
+
+        # Reject negative values (only happens outside of the brain)
+        noise[noise < 0] = 0
+
+    # Failed to converge
+    if iterations == 0:
+        logger.info('No fitting iterations were run')
+    elif iteration == iterations:
+        logger.info('SNR failed to converge.')
```

Comment:
I think that when iterations failed to converge this should through a warning but I don't think a warning is useful when iterations is set (by the user) to zero, although it would be strange to call the 'fit*' functions and not specify iterations.